### DISCUSSION TITLE
cross-compiling fails

### DISCUSSION THREAD

**Discussion by saruvora** [avrdudes/avrdude#1535] [Q&A]
Hi I have changed some thing in the CMakeLists.txt for cross-compiling for arm-linux-gnueabihf. 
My PC is running debian and arm as well. 

```
set(CMAKE_INSTALL_PREFIX "/path/to/arm")
# Specify the cross-compiler
set(CMAKE_C_COMPILER arm-linux-gnueabihf-gcc-10)

```
and then I run: 

`cmake -D HAVE_LINUXGPIO=1 -D HAVE_LIBELF=1 -D HAVE_LIBUSB=1 -D HAVE_LIBHIDAPI=0 -D HAVE_LIBFTDI1=1 -D HAVE_LIBUSB_1_0=1 -D HAVE_LIBREADLINE=0  -B build_arm
`

I run into following errors I have already install libusb-1-0.0-dev : 

```
Determining if the include file libelf/libelf.h exists failed with the following output:
Change Dir: /home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp
Run Build Command(s):/usr/bin/gmake cmTC_80844/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_80844.dir/build.make CMakeFiles/cmTC_80844.dir/build
gmake[1]: Entering directory '/home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_80844.dir/CheckIncludeFile.c.o
/usr/bin/cc --sysroot=/../   -std=gnu11 -o CMakeFiles/cmTC_80844.dir/CheckIncludeFile.c.o -c /home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp/CheckIncludeFile.c
/home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp/CheckIncludeFile.c:1:10: fatal error: libelf/libelf.h: No such file or directory
    1 | #include <libelf/libelf.h>

      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
gmake[1]: *** [CMakeFiles/cmTC_80844.dir/build.make:85: CMakeFiles/cmTC_80844.dir/CheckIncludeFile.c.o] Error 1
gmake[1]: Leaving directory '/home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp'
gmake: *** [Makefile:140: cmTC_80844/fast] Error 2
Determining if the include file lusb0_usb.h exists failed with the following output:
Change Dir: /home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp
Run Build Command(s):/usr/bin/gmake cmTC_fb0e5/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_fb0e5.dir/build.make CMakeFiles/cmTC_fb0e5.dir/build
gmake[1]: Entering directory '/home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_fb0e5.dir/CheckIncludeFile.c.o
/usr/bin/cc --sysroot=/../   -std=gnu11 -o CMakeFiles/cmTC_fb0e5.dir/CheckIncludeFile.c.o -c /home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp/CheckIncludeFile.c
/home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp/CheckIncludeFile.c:1:10: fatal error: lusb0_usb.h: No such file or directory
    1 | #include <lusb0_usb.h>
      |          ^~~~~~~~~~~~~
compilation terminated.
gmake[1]: *** [CMakeFiles/cmTC_fb0e5.dir/build.make:85: CMakeFiles/cmTC_fb0e5.dir/CheckIncludeFile.c.o] Error 1
gmake[1]: Leaving directory '/home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp'
gmake: *** [Makefile:140: cmTC_fb0e5/fast] Error 2
Determining if the ftdi_tcioflush exist failed with the following output:
Change Dir: /home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp
Run Build Command(s):/usr/bin/gmake cmTC_968ef/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_968ef.dir/build.make CMakeFiles/cmTC_968ef.dir/build
gmake[1]: Entering directory '/home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_968ef.dir/CheckSymbolExists.c.o
/usr/bin/cc --sysroot=/../   -std=gnu11 -o CMakeFiles/cmTC_968ef.dir/CheckSymbolExists.c.o -c /home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp/CheckSymbolExists.c
gmake[1]: *** No rule to make target '/home/developer/avrdude-7.2_/1', needed by 'cmTC_968ef'.  Stop.
gmake[1]: Leaving directory '/home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp'
gmake: *** [Makefile:140: cmTC_968ef/fast] Error 2
File /home/developer/avrdude-7.2_/build_arm/CMakeFiles/CMakeTmp/CheckSymbolExists.c:
/* */
#include <libftdi1/ftdi.h>
int main(int argc, char** argv)
{
 (void)argv;
#ifndef ftdi_tcioflush
  return ((int*)(&ftdi_tcioflush))[argc];
#else
  (void)argc;
  return 0;
#endif
}

```

Can someone help me resolve this issue, please? 
Also, is it possible to flash Atxmega256A3u which is connected to the armv7l via UARTD0?  without any extra hardware except USB.

**Comment by mcuee** [comment]
The wiki does not cover cross-compiling now. But you need to install more dependancies first.
https://github.com/avrdudes/avrdude/wiki/Building-AVRDUDE-for-Linux
(libserialport is not necessary for 7.2 release but needed for git main and the upcoming 7.3 release).

**Reply by saruvora** [reply]
Hi I have installed all these dependencies on the Host-PC.

I have also done this: 

```
export C_INCLUDE_PATH=/usr/include/libftdi1/:$C_INCLUDE_PATH
export C_INCLUDE_PATH=/usr/include/libusb/:$C_INCLUDE_PATH
```

**Reply by mcuee** [reply]
That is not correct. You need to install the dependancies for the cross-compiler.

**Reply by saruvora** [reply]
I am relatively new to this field. What kind of more dependencies should I install? And do I have to install them on my arm controller as well ? or just the Host pc and then the cmake with cross-compiling will take care of it?

**Reply by mcuee** [reply]
I am not an expert in CMake, hopefully others can help you better. But you can also ask in other places as it is not specific to avrdude.

You may want to read this to see if that helps..
https://cmake.org/cmake/help/book/mastering-cmake/chapter/Cross%20Compiling%20With%20CMake.html

**Reply by mcuee** [reply]
What is your ARM Controller board? If it is powerful enough, better build avrdude from your ARM controller board.

**Reply by saruvora** [reply]
yes, It is a custom one with IMX6UL. Which has 512MB of memory on it.
Thank you for the cmake link I will check it out. 

I have also been able to avrdude without USB on it. I tested it with "-c dryrun" seems to work but I want to flash the Atxmega256a3u with it. For that I need USB support.

**Reply by saruvora** [reply]
Is there a plan to update the documentation for Cross-compiling for avrdude?

**Comment by mcuee** [comment]
> Is there a plan to update the documentation for Cross-compiling for avrdude?

If you figure that out, you are welcome to contribute the cross-compiling documentation. We will update the Wiki.

It could be that you also need to change the CMake scripts file.

We do not have plan to work on this now as we do not have CMake experts now. We need the help from the community.

**Reply by saruvora** [reply]
ah okay I will let you know if I succeed then I will be happy to contribute.

**Comment by mcuee** [comment]
BTW, the script here is using cross-compile:
https://github.com/avrdudes/avrdude/blob/main/.github/workflows/arduino_packing.yml

It relies on the container here:
https://github.com/arduino/crossbuild

You may want to consult the above.

**Reply by saruvora** [reply]
Thank you so much for this. I will try this in the upcoming days.

**Comment by stefanrueger** [comment]
Closing the discussion as it's gone stale
